server/routes/user_router.go

```diff
@@ -0,0 +1,358 @@
+	err := json.NewDecoder(r.Body).Decode(&signUp)
+	if err != nil {
+		router.WriteErrResponse(w, err)
+		return
+	}
+
+	// validate mail
+	bool := internal.ValidateMail(signUp.Email)
+	if !bool {
+		router.WriteErrResponse(w, fmt.Errorf("email isn't valid %v", err))
+		return
+	}
+
+	u := models.User{
+		Name:     signUp.Name,
+		Email:    signUp.Email,
+		Password: signUp.Password,
+	}
+
+	// check if user already exists
+	_, err = router.db.GetUserByEmail(u.Email, router.secret)
+	if err == nil {
+		router.WriteMsgResponse(w, fmt.Errorf("user already exists"))
+	}
+
+	// password should be ACII , min 5 , max 10
+	validator := password.NewValidator(true, 5, 10)
+	err = validator.ValidatePassword(signUp.Password)
+	if err != nil {
+		router.WriteErrResponse(w, fmt.Errorf("error: %v password isn't valid", err))
+		return
+	}
+
+	// password and confirm password should match
+	if signUp.Password != signUp.ConfirmPassword {
+		router.WriteErrResponse(w, fmt.Errorf("password and confirm password don't match"))
+		return
+	}
+
+	// hash password
+	hashedPassword, err := internal.HashPassword(u.Password)
+	if err != nil {
+		router.WriteErrResponse(w, err)
+		return
+	}
+	u.Password = hashedPassword
+
+	// send verification code
+	code, err := internal.SendMail(u.Email)
+	if err != nil {
+		router.WriteErrResponse(w, err)
+		return
+	}
+	u.Code = code
+	fmt.Printf("code: %v\n", code)
+	msg := "Verfification Code has been sent to " + u.Email
+	router.db.SetCache(u.Email, u)
+	router.WriteMsgResponse(w, msg)
+
+}
+
+// get verification code to create user
+func (router *Router) VerifyUser(w http.ResponseWriter, r *http.Request) {
+	data := AuthDataInput{}
+	err := json.NewDecoder(r.Body).Decode(&data)
+	if err != nil {
+		router.WriteErrResponse(w, err)
+		return
+	}
+
+	cachedUser, err := router.db.GetCache(data.Email)
+	if err != nil {
+		router.WriteErrResponse(w, err)
+		return
+	}
+
+	if cachedUser.Code != data.Code {
+		router.WriteErrResponse(w, fmt.Errorf("wrong code"))
+		return
+	}
+
+	u, err := router.db.CreateUser(&cachedUser)
+	if err != nil {
+		router.WriteErrResponse(w, err)
+		return
+	}
+	router.WriteMsgResponse(w, u)
+}
+
+func (router *Router) SignInHandler(w http.ResponseWriter, r *http.Request) {
+	u := models.User{}
+	err := json.NewDecoder(r.Body).Decode(&u)
+	if err != nil {
+		router.WriteErrResponse(w, err)
+		return
+	}
+
+	user, err := router.db.GetUserByEmail(u.Email, router.secret)
+	if err != nil {
+		router.WriteErrResponse(w, err)
+	}
+
+	err = internal.VerifyPassword(user.Password, u.Password)
+	if err != nil {
+		router.WriteErrResponse(w, fmt.Errorf("error %v, Password is not correct", err))
+		return
+	}
+
+	token, err := internal.CreateJWT(&u, router.secret)
+	if err != nil {
+		router.WriteErrResponse(w, err)
+		return
+	}
+
+	if err != nil {
+		router.WriteErrResponse(w, err)
+		return
+	}
+
+	json.NewEncoder(w).Encode("Token :" + token)
+}
+
+func (router *Router) Home(w http.ResponseWriter, r *http.Request) {
```

Comment:
I think it will be a getUserHandler instead of home and we will return user

Reply:
there's function to return user already, what will be the difference ?